esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -1811,241 +1819,470 @@ void doWUQueryWithSort(IEspContext &context, IEspWUQueryRequest & req, IEspWUQue
-                        archivedWUs.append(*info.getClear());
-                    }
-                    archivedWUs.sort(compareWuids);
+        for (unsigned i=0; i<numberOfWUsReturned; i++)
+        {
+            const char *csline = cmd->queryId(i);
+            if (!csline || !*csline)
+                continue;
 
-                    archivedWuCache.add(filterStr, "AddWhenAvailable", hasMoreWU, numberOfWUsReturned, archivedWUs);
-                }
+            StringArray wuDataArray;
+            wuDataArray.appendList(csline, ",");
+
+            const char* wuid = wuDataArray.item(0);
+            if (isEmpty(wuid))
+            {
+                WARNLOG("Empty WUID in SCA_LIST response"); // JCS->KW - have u ever seen this happen?
+                continue;
             }
-            return;
-        };
+            const char* owner = wuDataArray.item(1);
+            bool canAccess = chooseWuAccessFlagsByOwnership(context.queryUserId(), owner, accessOwn, accessOthers) >= SecAccess_Read;
+            if (!lightWeight)
+            {
+                Owned<IEspECLWorkunit> info = createArchivedWUEntry(wuDataArray, canAccess);
+                archivedWUs.append(*info.getClear());
+            }
+            else
+            {
+                Owned<IEspECLWorkunitLW> info = createArchivedLWWUEntry(wuDataArray, canAccess);
+                archivedLWWUs.append(*info.getClear());
+            }
+        }
+        if (!lightWeight)
+            archivedWUs.sort(compareWuids);
+        else
+            archivedLWWUs.sort(compareLWWuids);
 
-        bool getHasMoreWU() { return hasMoreWU; };
-        unsigned getNumberOfWUsReturned() { return numberOfWUsReturned; };
+        archivedWuCache.add(filterStr, "AddWhenAvailable", hasMoreWU, numberOfWUsReturned, archivedWUs, archivedLWWUs);
+        return;
     };
 
+    bool getHasMoreWU() { return hasMoreWU; };
+    unsigned getNumberOfWUsReturned() { return numberOfWUsReturned; };
+};
+
+void doWUQueryFromArchive(IEspContext &context, const char* sashaServerIP, unsigned sashaServerPort,
+       ArchivedWuCache &archivedWuCache, unsigned cacheMinutes, IEspWUQueryRequest & req, IEspWUQueryResponse & resp)
+{
     unsigned pageStart = (unsigned) req.getPageStartFrom();
     unsigned pageSize = (unsigned) req.getPageSize();
     if(pageSize < 1)
         pageSize=500;
-    IArrayOf<IEspECLWorkunit> archivedWUs;
     Owned<IArchivedWUsReader> archiveWUsReader = new CArchivedWUsReader(context, sashaServerIP, sashaServerPort, archivedWuCache,
-        cacheMinutes, pageStart, pageSize, req);
-    archiveWUsReader->getArchivedWUs(archivedWUs);
+        cacheMinutes, pageStart, pageSize);
+
+    IArrayOf<IEspECLWorkunit> archivedWUs;
+    IArrayOf<IEspECLWorkunitLW> dummyWUs;
+    Owned<CWULightWeightQueryRequest> dummyReq = new CWULightWeightQueryRequest("WsWorkunits");
+    archiveWUsReader->getArchivedWUs(false, req, *dummyReq, archivedWUs, dummyWUs);
```

Comment:
I might be missing what distinguishes them, but sharing the cache like this while only storing either lightweight or regular workunits, what happens if I make the exact same query for archived workunits twice, but once using lightweight, then regular query?  Seems like the second would return an empty list?  Because it found the cache, but lightweight were stored but not regular?